pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/SourceIdentifierScanner.scala

```diff
@@ -0,0 +1,49 @@
+package uk.ac.wellcome.platform.idminter.steps
+
+import grizzled.slf4j.Logging
+import io.circe.Json
+import io.circe.optics.JsonPath.root
+import uk.ac.wellcome.json.JsonUtil.fromJson
+import uk.ac.wellcome.models.Implicits._
+import uk.ac.wellcome.models.work.internal.SourceIdentifier
+
+import scala.util.{Failure, Success, Try}
+
+object SourceIdentifierScanner extends Logging {
+  def scan(inputJson: Json): Try[List[SourceIdentifier]] =
+    Try(
+      iterate(
+        root.each.json.getAll(inputJson),
+        findIdentifier(root.sourceIdentifier.json.getOption(inputJson)).toList))
+
+  private def iterate(
```

Comment:
I wonder if there's a way to make this tail recursive as the stack is going to get huuuuge for big documents

Reply:
yeh,  I couldn't find a way to do that. I'll poke around a bit more but we might need to give it more memory